### PR TITLE
Updated New-ModuleDeployment.ps1

### PR DESCRIPTION
**Pull request by ahmadabdalla**
# Change

Added a custom action to deploy templates across different scopes

## Type of Change

- [x ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

**Comment by MrMCake** [comment]
Small suggestion: 

By fetching the Schema from the ARM template you could automatically determine the scope. One param less to cover ;) As a reference, check line 95 in this [script](https://github.com/MrMCake/Modules/blob/main/.github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1) out. 

Also I'd recommend to move the content of the step into a dedicated script as it makes testing easier and allows you to re-use the code for other purposes.

**Comment by ahmadabdalla** [comment]
@MrMCake ready for review

**Comment by ahmadabdalla** [comment]
@MrMCake @segraef 

Updated the following: New-ModuleDeployment.ps1 

- Added 'additionalTags' as an optional input. Allows injecting info from the pipeline as tags (i.e. commit id). 
- Support for multiple files as input (comma separated).
- Support for directory-based path (filters for .json files). 
- Support for retry (2 times - 5 seconds apart).
- Moved the 'removeDeployment' logic before the deployment.

Questions:
- Do we need the 'moduleName' parameter? 
- Do we need the 'removeDeployment' parameter?

**Review by MrMCake — CHANGES_REQUESTED** [review]

File: .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1 [hunk 1]
```diff
@@ -66,97 +69,123 @@ function New-ModuleDeployment {
...
+            Write-Verbose "Detected Parameter File(s)/Directory - Count: `n $($parameterFilePath.Count)`n"
         }
-        if ($removeDeployment) {
-            # Fetch tags of parameter file if any (- required for the remove process. Tags may need to be compliant with potential customer requirements)
-            $parameterFileTags = (ConvertFrom-Json (Get-Content -Raw -Path $parameterFilePath) -AsHashtable).parameters.tags.value
-            if (-not $parameterFileTags) {
-                $parameterFileTags = @{}
+
+        ## Iterate through each file
+        foreach ($file in $parameterFilePath) {
```
[the remark this answers is not in the document]
Comment (ahmadabdalla) [reply .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1]: Addressed

File: .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1 [hunk 2]
```diff
@@ -66,97 +69,123 @@ function New-ModuleDeployment {
...
-    }
-    
+                    else {
+                        Write-Verbose "Resource deployment Failed.. ($Retrycount/1) Retrying in 5 Seconds.. `n"
+                        Start-Sleep -Seconds 5
+                        $Retrycount = $Retrycount + 1
+                    }
+                } #end catch
+            } #end do
+            while ($Stoploop -eq $false) {
```
[the remark this answers is not in the document]
Comment (ahmadabdalla) [reply .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1]: Added upper limit as an (or) and added optional parameter, default is 3

File: .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1 [hunk 3]
```diff
@@ -66,97 +69,123 @@ function New-ModuleDeployment {
...
-        if ($removeDeployment) {
-            # Fetch tags of parameter file if any (- required for the remove process. Tags may need to be compliant with potential customer requirements)
-            $parameterFileTags = (ConvertFrom-Json (Get-Content -Raw -Path $parameterFilePath) -AsHashtable).parameters.tags.value
-            if (-not $parameterFileTags) {
-                $parameterFileTags = @{}
+
+        ## Iterate through each file
+        foreach ($file in $parameterFilePath) {
+            $fileProperties = Get-Item -Path $file
+            Write-Verbose "Deploying: $($fileProperties.Name) `n"
```
[the remark this answers is not in the document]
Comment (ahmadabdalla) [reply .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1]: Resolved

**Review by MrMCake — APPROVED** [review]

File: .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (MrMCake) [reply .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1]: Approved

File: .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1 (hunk 1 above)
[the remark this answers is not in the document]
Comment (MrMCake) [reply .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1]: Approved

File: .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (MrMCake) [reply .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1]: Approved

**Pull request merged**